topdown/crypto.go

```diff
@@ -205,6 +209,55 @@ func builtinCryptoX509ParseRSAPrivateKey(_ BuiltinContext, operands []*ast.Term,
 	return iter(ast.NewTerm(value))
 }
 
+func builtinCryptoParsePrivateKey(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+
+	a := operands[0].Value
+	input, err := builtins.StringOperand(a, 1)
+	if err != nil {
+		return err
+	}
+
+	// get the raw private key
+	rawKey, err := getRSAPrivateKeyFromString(string(input))
+	if err != nil {
+		return err
+	}
+	key, err := parsePrivateKey(rawKey)
+
+	theJsonKey, err := json.Marshal(key)
+	if err != nil {
+		return err
+	}
+
+	var x interface{}
+	if err := util.UnmarshalJSON(theJsonKey, &x); err != nil {
+		return err
+	}
+
+	value, err := ast.InterfaceToValue(x)
+	if err != nil {
+		return err
+	}
+
+	return iter(ast.NewTerm(value))
+}
+
+func parsePrivateKey(privatekey interface{}) (interface{}, error) {
```

Comment:
thought: do we need to support DSA keys here too? the ssh package does so here: https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.11.0:ssh/keys.go;l=1111 but I'm not sure how important it is.

Reply:
from godoc: [Deprecated: DSA is a legacy algorithm, and modern alternatives such as Ed25519 (implemented by package crypto/ed25519) should be used instead. Keys with 1024-bit moduli (L1024N160 parameters) are cryptographically weak, while bigger keys are not widely supported. Note that FIPS 186-5 no longer approves DSA for signature generation](https://pkg.go.dev/crypto/dsa).


I wrote a quick test for PKCS8 ED25519, and it is passing. I'll add it for completion.